### PR TITLE
Note for E0599 if shadowed bindings has the method.

### PR DESCRIPTION
**Pull request by surechen**
implement #123558

Use a visitor to find earlier shadowed bingings which has the method.

r? @estebank

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 1]
```diff
@@ -358,9 +360,190 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                        self.tcx.hir().span(recv_id),
+                        format!(
+                            "earlier `{rcvr_name}` shadowed here with type `{ty_str_reported}`"
+                        ),
+                    );
+                    err.span_note(
+                        span,
+                        format!(
+                            "there's an earlier shadowed binding `{rcvr_name}` of type `{self_ty}` \
+                                    that *has* method `{method_name}` available"
```
Comment (estebank) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:535]:
```suggestion
                            "there's an earlier shadowed binding `{rcvr_name}` of type `{self_ty}` \
                              that has method `{method_name}` available"
```

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 2]
```diff
@@ -358,9 +360,190 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                if let Some(sub_var_scope) = scope_tree.var_scope(sub_id)
+                    && let Some(super_var_scope) = scope_tree.var_scope(super_id)
+                    && scope_tree.is_subscope_of(sub_var_scope, super_var_scope)
+                {
+                    return true;
+                }
+                false
+            }
+
+            fn check_and_add_sugg_binding(&mut self, binding: LetStmt) -> bool {
```
Comment (estebank) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:412]: More detailed comments inside of this method would be useful to allow people to skip the method with more ease.

**Review by estebank — APPROVED** [review]
My apologies for the delay reviewing.

r=me after rebasing

File: compiler/rustc_hir_typeck/src/method/suggest.rs (hunk 2 above)
Comment (surechen) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:412]: > More detailed comments inside of this method would be useful to allow people to skip the method with more ease.

Done. Thank you.

File: compiler/rustc_hir_typeck/src/method/suggest.rs (hunk 1 above)
Comment (surechen) [reply compiler/rustc_hir_typeck/src/method/suggest.rs]: > 

Done. Thank you.

**Comment by surechen** [comment]
> My apologies for the delay reviewing.
> 
> r=me after rebasing


Thank you very much. Please help me review again.

**Comment by estebank** [comment]
There are a few typos in the comments and formatting issues that `./x.py fmt` didn't seem to pick up/complain about, but if that's the case it's fine to land and clean up in a bit.

@bors r+

**Pull request merged**